document/core/binary/types.rst

```diff
@@ -90,14 +90,21 @@ Limits
 Memory Types
 ~~~~~~~~~~~~
 
-:ref:`Memory types <syntax-memtype>` are encoded with their :ref:`limits <binary-limits>`.
+:ref:`Memory types <syntax-memtype>` are encoded with their :ref:`limits <binary-limits>` that includes an extra value to specify whether the the memory is shared.
 
 .. math::
    \begin{array}{llclll@{\qquad\qquad}l}
    \production{memory type} & \Bmemtype &::=&
-     \X{lim}{:}\Blimits &\Rightarrow& \X{lim} \\
+     \hex{00}~~n{:}\Bu32 &\Rightarrow& \{ \LMIN~n, \LMAX~\epsilon \}~\MNOTSHARED \\ &&|&
```

Comment:
Instead of this "code duplication", can the binary grammar for limits perhaps return an additional flags byte representing the other bits of the opcode? Then you just need to enforce that to be zero in the table case. (Potentially, table types might grow some flags as well in the future, so that it could even be useful to make it reusable.)

Reply:
I like this idea, but I'm not sure quite how to represent this in the doc. It seems that most of the "functions" (such as `limits` above) return a specific syntax type, not an arbitrary struct or pair of values. Also, how can I extract multiple values at the callsite (such as in `memtype` or `tabletype`)?

Reply:
There are a couple of other rules that synthesise a pair, e.g. [type uses](https://webassembly.github.io/spec/text/modules.html#type-uses) in the text format. At the use site I just pattern-match the pair, see e.g. [import descriptions](https://webassembly.github.io/spec/text/modules.html#text-importdesc).

Reply:
Ah, that makes sense. Thanks!